File: backend/api/v1/data_sources.py
"""Data source API router for PolyEdge plugin system."""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.api.auth import require_admin
from backend.data.source_registry import source_registry
from backend.core.plugin_errors import PluginNotFound

router = APIRouter(tags=["Data Sources"])


@router.get("/sources")
async def list_sources(_: Session = Depends(require_admin)):
    """List all available data sources."""
    return {
        "sources": [
            {
                "name": m.name,
                "display_name": m.display_name,
                "version": m.version,
                "data_types": [dt.value for dt in m.data_types],
                "supports_streaming": m.supports_streaming,
                "supports_backfill": m.supports_backfill,
                "is_live": m.is_live,
                "rate_limit_per_minute": m.rate_limit_per_minute,
                "tags": m.tags,
            }
            for m in source_registry.list_available()
        ]
    }


@router.get("/sources/{name}")
async def get_source(name: str, _: Session = Depends(require_admin)):
    """Get details for a specific data source."""
    try:
        source = source_registry.get(name)
        manifest = source.manifest()
        return {
            "name": manifest.name,
            "display_name": manifest.display_name,
            "version": manifest.version,
            "data_types": [dt.value for dt in manifest.data_types],
            "supports_streaming": manifest.supports_streaming,
            "supports_backfill": manifest.supports_backfill,
            "is_live": manifest.is_live,
            "required_env_vars": manifest.required_env_vars,
            "enabled": source_registry._enabled.get(name, False),
            "healthy": source_registry._health_status.get(name, False),
        }
    except PluginNotFound as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/sources/{name}/enable")
async def enable_source(name: str, _: Session = Depends(require_admin)):
    """Enable a data source."""
    try:
        source_registry.set_enabled(name, True)
        return {"status": "enabled", "name": name}
    except PluginNotFound as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/sources/{name}/disable")
async def disable_source(name: str, _: Session = Depends(require_admin)):
    """Disable a data source."""
    try:
        source_registry.set_enabled(name, False)
        return {"status": "disabled", "name": name}
    except PluginNotFound as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/src/api/data_sources.ts
import { api } from '../api'

export interface DataSource {
  name: string
  display_name: string
  version: string
  data_types: string[]
  supports_streaming: boolean
  supports_backfill: boolean
  is_live: boolean
  rate_limit_per_minute: number
  tags: string[]
}

export interface DataSourceDetails extends DataSource {
  description?: string
  required_env_vars: string[]
  enabled: boolean
  healthy: boolean
}

export interface MarketVenue {
  name: string
  display_name: string
  version: string
  venue_type: string
  capabilities: string[]
  supported_currencies: string[]
  is_live_venue: boolean
  supports_paper_mode: boolean
  min_order_size_usd: number
  maker_fee_bps: number
  taker_fee_bps: number
  tags: string[]
}

export interface MarketVenueDetails extends MarketVenue {
  required_env_vars: string[]
  enabled: boolean
  healthy: boolean
}

export interface VenueBalance {
  total_balance_usd: number
  available_balance_usd: number
  wallet_address?: string
  last_updated: string
}

export interface VenuePosition {
  market_id: string
  market_question?: string
  position_size: number
  entry_price: number
  current_price?: number
  pnl?: number
  status: string
}

export const dataSourcesAPI = {
  list: () => api.get<{ sources: DataSource[] }>('/sources').then(r => r.data.sources),
  get: (name: string) => api.get<DataSourceDetails>(`/sources/${name}`).then(r => r.data),
  enable: (name: string) => api.post<{ status: string; name: string }>(`/sources/${name}/enable`).then(r => r.data),
  disable: (name: string) => api.post<{ status: string; name: string }>(`/sources/${name}/disable`).then(r => r.data),
}